src/data/matrix/hadamard.lean

```diff
@@ -0,0 +1,142 @@
+/-
+Copyright (c) 2021 Lu-Ming Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Lu-Ming Zhang
+-/
+import data.matrix.notation
+import linear_algebra.matrix.trace
+import data.complex.basic
+
+/-!
+# Hadamard product of matrices
+
+This file defines the Hadamard product `matrix.Hadamard`
+and contains basic properties about them.
+
+## Main definition
+
+- `matrix.Hadamard`: defines the Hadamard product,
+  which is the pointwise product of two matrices of the same size.
+
+## Notation
+
+* `⊙`: the Hadamard product `matrix.Hadamard`;
+
+## References
+
+*  <https://en.wikipedia.org/wiki/Hadamard_product_(matrices)>
+
+## Tags
+
+Hadamard product, Hadamard
+-/
+
+variables {α β γ I J K L M N: Type*}
+variables {R : Type*}
+variables {m n p q r s t: ℕ}
+
+namespace matrix
+open_locale matrix big_operators
+open complex
+
+/-- `matrix.Hadamard` defines the Hadamard product,
+    which is the pointwise product of two matrices of the same size.-/
+def Hadamard [has_mul α] (A : matrix I J α) (B : matrix I J α) :
+matrix I J α :=
+λ i j, (A i j) * (B i j)
+
+localized "infix `⊙`:100 := matrix.Hadamard" in matrix -- declares the notation
+
+section basic_properties
+
+variables (A : matrix I J α) (B : matrix I J α) (C : matrix I J α)
+
+/- commutativity -/
+lemma Hadamard_comm [comm_semigroup α] : A ⊙ B = B ⊙ A :=
+by ext; simp [Hadamard, mul_comm]
+
+/- associativity -/
+lemma Hadamard_assoc [semigroup α] : A ⊙ B ⊙ C = A ⊙ (B ⊙ C) :=
+by ext; simp [Hadamard, mul_assoc]
+
+/- distributivity -/
+lemma Hadamard_add [distrib α] : A ⊙ (B + C) = A ⊙ B + A ⊙ C :=
+by ext; simp [Hadamard, left_distrib]
+
+lemma add_Hadamard [distrib α] : (B + C) ⊙ A = B ⊙ A + C ⊙ A :=
+by ext; simp [Hadamard, right_distrib]
+
+
+/- scalar multiplication -/
+section scalar
+
+@[simp] lemma smul_Hadamard
+[has_mul α] [has_scalar R α] [is_scalar_tower R α α] (k : R) :
+(k • A) ⊙ B = k • A ⊙ B :=
+by {ext, simp [Hadamard], exact smul_assoc _ (A i j) _}
+
+@[simp] lemma Hadamard_smul
+[has_mul α] [has_scalar R α] [smul_comm_class R α α] (k : R):
+A ⊙ (k • B) = k • A ⊙ B :=
+by {ext, simp [Hadamard], exact (smul_comm k (A i j) (B i j)).symm}
+
+end scalar
+
+section zero
+
+variables [mul_zero_class α]
+
+@[simp] lemma Hadamard_zero : A ⊙ (0 : matrix I J α) = 0 :=
+by ext; simp [Hadamard]
+
+@[simp] lemma zero_Hadamard : (0 : matrix I J α) ⊙ A = 0 :=
+by ext; simp [Hadamard]
+
+end zero
+
+section trace
+
+variables [fintype I] [fintype J]
+variables [semiring β] [semiring R] [comm_semiring α] [module β α] [module R ℂ]
+
+lemma trace_identity [decidable_eq I] [decidable_eq J] (v : I → α) (w : J → α):
+dot_product (vec_mul  v  (A ⊙ B)) w =
+trace I β α ((diagonal v)ᵀ ⬝ A ⬝ (diagonal w) ⬝ Bᵀ) :=
```

Comment:
If you change the statement as follows then you don't need `α` to be commtative any more
```suggestion
trace I β α ((diagonal v)ᵀ ⬝ A ⬝ (B ⬝ diagonal w)ᵀ) :=
```

Reply:
I personally prefer the current statement as it looks like a more natural math formula to me.

Reply:
The first transpose is redundant actually. I will reformulate it.

Reply:
> as it looks like a more natural math formula to me.

Sure, but that's only because commutative rings are more natural than non-commutative ones! I think it's worth tweaking the statement to not require commutativity; the user can always rewrite by `matrix.transpose_mul` to clean things up after applying this lemma if their ring _is_ commutative.

Reply:
" commutative rings are more natural than non-commutative ones!" good point